src/primitives/hrp.rs

```diff
@@ -0,0 +1,420 @@
+//! > character having a value in the range [33-126]. HRP validity may be further restricted by
+//! > specific applications.
+//!
+//! ref: [BIP-173](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki#user-content-Bech32)
+
+#[cfg(feature = "alloc")]
+use alloc::string::String;
+use core::cmp::Ordering;
+use core::fmt::{self, Write};
+use core::iter::FusedIterator;
+use core::slice::Iter;
+
+use crate::Case;
+
+/// Maximum length of the human-readable part, as defined by BIP-173.
+const MAX_HRP_LEN: usize = 83;
+
+/// The human-readable part (human readable prefix before the '1' separator).
+#[derive(Clone, Copy, Debug)]
+pub struct Hrp {
+    /// ASCII byte values, guaranteed not to be mixed-case.
+    buf: [u8; MAX_HRP_LEN],
+    /// Number of characters currently stored in this HRP.
+    size: usize,
+}
+
+impl Hrp {
+    /// Parses the human-readable part checking it is valid as defined by [BIP-173].
+    ///
+    /// This does _not_ check that the `hrp` is an in-use HRP within Bitcoin (eg, "bc"), rather it
+    /// checks that the HRP string is valid as per the specification in [BIP-173]:
+    ///
+    /// > The human-readable part, which is intended to convey the type of data, or anything else that
+    /// > is relevant to the reader. This part MUST contain 1 to 83 US-ASCII characters, with each
+    /// > character having a value in the range [33-126]. HRP validity may be further restricted by
+    /// > specific applications.
+    ///
+    /// [BIP-173]: <https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki>
+    pub fn parse(hrp: &str) -> Result<Self, Error> { Self::parse_and_case(hrp).map(|(p, _)| p) }
+
+    /// Returns the case as well as the parsed `Hrp`.
+    pub(crate) fn parse_and_case(hrp: &str) -> Result<(Self, Case), Error> {
+        use Error::*;
+
+        let mut new = Hrp { buf: [0_u8; MAX_HRP_LEN], size: 0 };
+
+        if hrp.is_empty() {
+            return Err(Empty);
+        }
+        if hrp.len() > MAX_HRP_LEN {
+            return Err(TooLong(hrp.len()));
+        }
+
+        let mut has_lower: bool = false;
+        let mut has_upper: bool = false;
+        for (i, b) in hrp.bytes().enumerate() {
+            // Valid subset of ASCII
+            if !(33..=126).contains(&b) {
+                return Err(InvalidAsciiByte(b));
+            }
+
+            if b.is_ascii_lowercase() {
+                has_lower = true;
+            } else if b.is_ascii_uppercase() {
+                has_upper = true;
+            };
+
+            if has_lower && has_upper {
+                return Err(MixedCase);
+            }
+            new.buf[i] = b;
+            new.size += 1;
+        }
+
+        let case = match (has_lower, has_upper) {
+            (true, false) => Case::Lower,
+            (false, true) => Case::Upper,
+            (false, false) => Case::None,
+            (_, _) => unreachable!("checked above"),
```

Comment:
In 599ce6d089904386ac6a34d77a6b3b1fe222393f:

I think it would be nicer to just move the error return here. It means you lose the short-circuiting but it also means that you eliminate a panic branch.

Reply:
How about this so we keep the short-circuiting
```
            (_, _) => return Err(MixedCase), // Defensive programming, this statement is unreachable.
```

Reply:
Yeah, I'd be cool with that.